text/0000-async-stream.md

```diff
@@ -0,0 +1,903 @@
+
+    fn next(&mut self) -> Next<'_, Self>
+    where
+        Self: Unpin;
+}
+```
+
+Iterator does not require pinning its core next method. In order for a `gen fn` to operate with the Iterator ecosystem, there must be some kind of initial pinning step that converts its result into an iterator. This will be tricky, since you can't return a pinned value except by boxing. 
+
+The general shape will be:
+
+```rust
+gen_fn().pin_somehow().adapter1().adapter2()
+```
+
+With streams, the core interface _is_ pinned, so pinning occurs at the last moment.
+
+The general shape would be 
+
+```rust
+async_gen_fn().adapter1().adapter2().pin_somehow()
+```
+
+Pinning at the end, like with a stream, lets you build and return those adapters and then apply pinning at the end. This may be the more efficient setup and implies that, in order to have a `gen fn` that produces iterators, we will need to potentially disallow borrowing yields or implement some kind of `PinnedIterator` trait that can be "adapted" into an iterator by pinning.
+
+For example: 
+
+```rust
+trait PinIterator {
+    type Item;
+    fn next(self: Pin<&mut Self>) -> Self::Item;
+    // combinators can go here (duplicating Iterator for the most part)
+}
+impl<I: PinIterator, P: Deref<Target = I> + DerefMut> Iterator for Pin<P> {
+    type Item = <I as PinIterator>::Item;
+    fn next(&mut self) -> Self::Item { self.as_mut().next() }
+}
+
+// this would be nice.. but would lead to name resolution ambiguity for our combinators 😬 
+default impl<T: Iterator> PinIterator for T { .. }
+```
+
+Pinning also applies to the design of AsyncRead/AsyncWrite, which currently uses Pin even through there is no clear plan to make them implemented with generator type syntax. The asyncification of a signature is currently understood as pinned receiver + context arg + return poll.
+
+Another key difference between `Iterators` and `Streams` is that that futures are ultimately passed to some executor API like spawn which expects a static future. To achieve that, the futures contain all the state they need and references are internal to that state. Iterators are almost never required to be 'static by the APIs that consume them.
+
+It is, admittedly, somewhat confusing to have Async generators require Pinning and Iterator generators to not require pinning, users may feel they are creating code in an unnatural way when using the Async generators. This will need to be discussed more when generators are proposed in the future.
+
+### Yielding
+
+It would be useful to be able to yield from inside a for loop, as long as the for loop is
+over a borrowed input and not something owned by the stack frame.
+
+In the spirit of experimentation, boats has written the [propane](https://github.com/withoutboats/propane) 
+crate. This crate includes a `#[propane] fn` that changes the function signature
+to return `impl Iterator` and lets you `yield`. The non-async version uses 
+`static generator` that is currently in nightly only.
```

Comment:
This section actually looks to be about disallowing self-borrowing in non-async generators. That's the main design point propane is trying to prove, as I understand it.

cc @withoutboats to confirm, and also check my statement that yielding from a for loop over a borrowed input still works.

Suggested wording:

```suggestion
### Disallowing self-borrowing generators in `gen fn`

Another option is to make the generators returned by `gen fn` always be `Unpin` so that the user doesn't have to think about pinning unless they're already in an async context.

In the spirit of experimentation, boats has written the [propane](https://github.com/withoutboats/propane) 
crate. This crate includes a `#[propane] fn` that changes the function signature
to return `impl Iterator` and lets you `yield`. The non-async version uses 
nightly-only `static` generators, which disallow self-borrowing. In other words,
you can't hold a reference to something on the stack across a `yield`.

This should still allow yielding from inside a for loop, as long as the for loop is
over a borrowed input and not something owned by the stack frame.
```

Reply:
"`static` generators" are precisely the generators which _do_ support self-borrowing; `propane` uses non-`static` generators.

Reply:
whoops, I still get those mixed up, thanks. Edited the suggestion.

Reply:
@withoutboats when you have a moment can you confirm @tmandry's statement that yielding from a for loop over a borrowed input still works? Thank you!

Reply:
I find @tmandry's rewording accurate in describing propane :)